unified-runtime/source/adapters/opencl/event.hpp

```diff
@@ -0,0 +1,49 @@
+//===--------- queue.hpp - OpenCL Adapter ---------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include "common.hpp"
+
+#include <vector>
+
+struct ur_event_handle_t_ {
+  using native_type = cl_event;
+  native_type CLEvent;
+  ur_context_handle_t Context;
+  ur_queue_handle_t Queue;
+  std::atomic<uint32_t> RefCount = 0;
+  bool IsNativeHandleOwned = true;
+
+  ur_event_handle_t_(native_type Event, ur_context_handle_t Ctx,
+                     ur_queue_handle_t Queue)
+      : CLEvent(Event), Context(Ctx), Queue(Queue) {
+    RefCount = 1;
+    urContextRetain(Context);
+    if (Queue) {
```

Comment:
rather than allowing `Queue` to be null I think we should query it out of `Event` here, otherwise we're introducing a regression when you query EVENT_INFO_QUEUE for an event created from a native handle since the implementation for that query just returns `hEvent->Queue`

Reply:
actually that doesn't work because the query should return the corresponding ur handle not the native one, not sure what to do about this but it is a regression

Reply:
I've made it so that if you query `EVENT_INFO_QUEUE` on an event with no internal queue handle, it creates a new one on the fly and takes ownership of it.